### PR TITLE
[AutoDiff] Add 'Builtin.applyTranspose*'.

### PR DESCRIPTION
**Pull request by rxwei**
Add support for applying the transpose function in a `@differentiable(linear)` function. The `Builtin.applyTranspose*` builtin takes a `@differentiable(linear)` function and a tangent vector, and returns the result of applying the transpose to the tangent vector.

Pseudo-declaration:
```swift
func applyTranspose_arity{arity}[_throws?{r}]<T...{arity}, R>(
    _: @differentiable (T...) {r}?throws -> R,
    _: R
) {r}?rethrows -> (T...)
    where T: Differentiable & AdditiveArithmetic, R: Differentiable & AdditiveArithmetic
```

This patch also renames `Builtin.autodiffApply` to `Builtin.applyDerivative` for clarity, and fixes a bug in `LinearDifferentiableSILFunctionTypeLowering` where it expected 3 component values in `rebuildAggregate` instead of 2.

Resolves SR-11844 and SR-11851.

### CODE REVIEW

File: lib/SILGen/SILGenBuiltin.cpp [hunk 1]
```diff
@@ -1035,6 +1035,9 @@ static ManagedValue emitBuiltinAutoDiffApplyDerivativeFunction(
     AutoDiffDerivativeFunctionKind kind, unsigned arity,
     bool rethrows, SILGenFunction &SGF, SILLocation loc,
     SubstitutionMap substitutions, ArrayRef<ManagedValue> args, SGFContext C) {
+  // FIXME(SR-11853): Support throwing functions.
+  assert(!rethrows && "Thowing functions are not yet supported");
```
Comment (dan-zheng) [inline lib/SILGen/SILGenBuiltin.cpp]:
```suggestion
  assert(!rethrows && "Throwing functions are not yet supported");
```

File: lib/SILGen/SILGenBuiltin.cpp [hunk 2]
```diff
@@ -1133,11 +1136,45 @@ static ManagedValue emitBuiltinAutoDiffApplyDerivativeFunction(
...
-                                             SILLocation loc,
-                                             SubstitutionMap substitutions,
-                                             ArrayRef<ManagedValue> args,
-                                             SGFContext C) {
+// SWIFT_ENABLE_TENSORFLOW
+static ManagedValue emitBuiltinAutoDiffApplyTransposeFunction(
+    unsigned arity, bool rethrows, SILGenFunction &SGF, SILLocation loc,
+    SubstitutionMap substitutions, ArrayRef<ManagedValue> args, SGFContext C) {
+  // FIXME(SR-11853): Support throwing functions.
+  assert(!rethrows && "Thowing functions are not yet supported");
```
Comment (dan-zheng) [inline lib/SILGen/SILGenBuiltin.cpp]:
```suggestion
  assert(!rethrows && "Throwing functions are not yet supported");
```

File: lib/AST/AutoDiff.cpp [hunk 3]
```diff
@@ -206,21 +206,77 @@ void autodiff::getSubsetParameterTypes(IndexSubset *subset,
   }
 }
 
-bool autodiff::getBuiltinAutoDiffApplyConfig(
+// Given the rest of a `Builtin.applyDerivative_{jvp|vjp}` or
+// `Builtin.applyTranspose` operation name, attempts to parse the arity and
+// throwing-ness from the operation name. Modifies the operation name argument
```
Comment (dan-zheng) [inline lib/AST/AutoDiff.cpp]: minor:
```suggestion
// rethrows-ness from the operation name. Modifies the operation name argument
```

**Review by dan-zheng — APPROVED** [review]
Nice!

---

Note for self: `Builtin.apply{Derivative,Transpose}*` and `Builtin.{differentiable,linear}Function*` are all higher-order functions.

`apply{Derivative,Transpose}` are `rethrows` because they may throw depending on whether the argument function throws.

`{differentiable,linear}Function` are not throwing because they merely form a bundle of functions.

File: lib/AST/AutoDiff.cpp (hunk 3 above)
Comment (rxwei) [reply lib/AST/AutoDiff.cpp]: This has now been renamed to `throws`.

File: lib/SILGen/SILGenBuiltin.cpp (hunk 2 above)
Comment (rxwei) [reply lib/SILGen/SILGenBuiltin.cpp]: Done.

**Comment by rxwei** [comment]
> `apply{Derivative,Transpose}` are `rethrows` because they may throw depending on whether the argument function throws.

I just unified the terminology to just "throws" for simplicity :)

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Pull request merged**